kv/kvclient/kvcoord: TestTxnCoordSenderPipelining failed

kv/kvclient/kvcoord.TestTxnCoordSenderPipelining [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3301266&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3301266&tab=artifacts#/) on master @ [ed9411c488e93dca28a61a2abafc087aac00946e](https://github.com/cockroachdb/cockroach/commits/ed9411c488e93dca28a61a2abafc087aac00946e):


```
=== RUN   TestTxnCoordSenderPipelining
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnCoordSenderPipelining320077597
    test_log_scope.go:74: use -show-logs to present logs inline
    txn_coord_sender_test.go:2239: 
        	Error Trace:	txn_coord_sender_test.go:2239
        	Error:      	Not equal: 
        	            	expected: []roachpb.Method{1, 22, 10, 1, 10}
        	            	actual  : []roachpb.Method{1, 22, 10, 1, 17, 10}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-([]roachpb.Method) (len=5) {
        	            	+([]roachpb.Method) (len=6) {
        	            	  (roachpb.Method) 1,
        	            	@@ -5,2 +5,3 @@
        	            	  (roachpb.Method) 1,
        	            	+ (roachpb.Method) 17,
        	            	  (roachpb.Method) 10
        	Test:       	TestTxnCoordSenderPipelining
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnCoordSenderPipelining320077597
--- FAIL: TestTxnCoordSenderPipelining (1.41s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTxnCoordSenderPipelining PKG=./pkg/kv/kvclient/kvcoord TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
benesch marked as alumn{us/a}; resolving to nvanbenschoten instead
```
</p>
</details>

/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnCoordSenderPipelining.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestTxnCoordSenderPipelining [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3305594&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3305594&tab=artifacts#/) on master @ [04a41e7915f4a89dcc1d0dbd92466c6adf79ec9f](https://github.com/cockroachdb/cockroach/commits/04a41e7915f4a89dcc1d0dbd92466c6adf79ec9f):


```
=== RUN   TestTxnCoordSenderPipelining
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnCoordSenderPipelining669158218
    test_log_scope.go:74: use -show-logs to present logs inline
    txn_coord_sender_test.go:2239: 
        	Error Trace:	txn_coord_sender_test.go:2239
        	Error:      	Not equal: 
        	            	expected: []roachpb.Method{1, 22, 10, 1, 10}
        	            	actual  : []roachpb.Method{1, 22, 10, 1, 17, 10}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-([]roachpb.Method) (len=5) {
        	            	+([]roachpb.Method) (len=6) {
        	            	  (roachpb.Method) 1,
        	            	@@ -5,2 +5,3 @@
        	            	  (roachpb.Method) 1,
        	            	+ (roachpb.Method) 17,
        	            	  (roachpb.Method) 10
        	Test:       	TestTxnCoordSenderPipelining
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTxnCoordSenderPipelining669158218
--- FAIL: TestTxnCoordSenderPipelining (1.08s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTxnCoordSenderPipelining PKG=./pkg/kv/kvclient/kvcoord TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
benesch marked as alumn{us/a}; resolving to nvanbenschoten instead
```
</p>
</details>

/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnCoordSenderPipelining.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/txn_coord_sender_factory.go
+++ b/pkg/kv/kvclient/kvcoord/txn_coord_sender_factory.go
@@ -50,6 +50,7 @@ type TxnCoordSenderFactoryConfig struct {
 	Clock    *hlc.Clock
 	Stopper  *stop.Stopper
 
+	// -1 to disable transaction heartbeats.
 	HeartbeatInterval time.Duration
 	Linearizable      bool
 	Metrics           TxnMetrics

--- a/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go
@@ -2216,6 +2216,9 @@ func TestTxnCoordSenderPipelining(t *testing.T) {
 		Settings:   s.Cfg.Settings,
 		Clock:      s.Clock,
 		Stopper:    s.Stopper(),
+		// Disable transaction heartbeats so that they don't disrupt our attempt to
+		// track the requests issued by the transactions.
+		HeartbeatInterval: -1,
 	}, senderFn)
 	db := kv.NewDB(ambientCtx, tsf, s.Clock, s.Stopper())
 

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_heartbeater.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_heartbeater.go
@@ -282,6 +282,10 @@ func (h *txnHeartbeater) closeLocked() {
 
 // startHeartbeatLoopLocked starts a heartbeat loop in a different goroutine.
 func (h *txnHeartbeater) startHeartbeatLoopLocked(ctx context.Context) {
+	if h.loopInterval < 0 {
+		log.Infof(ctx, "coordinator heartbeat loop disabled")
+		return
+	}
 	if h.mu.loopStarted {
 		log.Fatal(ctx, "attempting to start a second heartbeat loop")
 	}

